cudax/include/cuda/experimental/__device/all_devices.cuh

```diff
@@ -0,0 +1,244 @@
+  {
+    return __proxy{device(__id_)};
+  }
+
+  constexpr iterator& operator++() noexcept
+  {
+    ++__id_;
+    return *this;
+  }
+
+  constexpr iterator operator++(int) noexcept
+  {
+    auto tmp = *this;
+    ++__id_;
+    return tmp;
+  }
+
+  constexpr iterator& operator--() noexcept
+  {
+    --__id_;
+    return *this;
+  }
+
+  constexpr iterator operator--(int) noexcept
+  {
+    auto tmp = *this;
+    --__id_;
+    return tmp;
+  }
+
+  constexpr iterator& operator+=(int __offset) noexcept
+  {
+    __id_ += __offset;
+    return *this;
+  }
+
+  _CCCL_NODISCARD constexpr iterator operator+(int __offset) noexcept
+  {
+    return iterator(__id_ + __offset);
+  }
+
+  constexpr iterator& operator-=(int __offset) noexcept
+  {
+    __id_ -= __offset;
+    return *this;
+  }
+
+  _CCCL_NODISCARD constexpr iterator operator-(int __offset) noexcept
+  {
+    return iterator(__id_ - __offset);
+  }
+
+  _CCCL_NODISCARD_FRIEND constexpr iterator operator+(int __lhs, const iterator& __rhs) noexcept
+  {
+    return iterator(__lhs + __rhs.__id_);
+  }
+
+  _CCCL_NODISCARD_FRIEND constexpr int operator-(const iterator& __lhs, const iterator& __rhs) noexcept
+  {
+    return __lhs.__id_ - __rhs.__id_;
+  }
+
+  _CCCL_NODISCARD_FRIEND constexpr bool operator==(const iterator& __lhs, const iterator& __rhs) noexcept
+  {
+    return __lhs.__id_ == __rhs.__id_;
+  }
+
+  _CCCL_NODISCARD_FRIEND constexpr bool operator!=(const iterator& __lhs, const iterator& __rhs) noexcept
+  {
+    return __lhs.__id_ != __rhs.__id_;
+  }
+
+private:
+  friend struct all_devices;
+
+  constexpr explicit iterator(int __id) noexcept
+      : __id_(__id)
+  {}
+
+  int __id_ = 0;
+};
+
+_CCCL_NODISCARD inline constexpr device all_devices::operator[](size_type __id_) const noexcept
+{
+  assert(__id_ < size());
+  return device(static_cast<int>(__id_));
+}
+
+_CCCL_NODISCARD inline all_devices::size_type all_devices::size() const
+{
+  size_type __size = __size_.load(_CUDA_VSTD::memory_order_relaxed);
+  if (__size == ~0UL)
+  {
+    int __count = 0;
+    _CCCL_TRY_CUDA_API(::cudaGetDeviceCount, "failed to get the count of CUDA devices", &__count);
+    __size = static_cast<size_type>(__count);
+    __size_.store(__size, _CUDA_VSTD::memory_order_relaxed);
+  }
+  return __size;
+}
+
+_CCCL_NODISCARD inline all_devices::iterator all_devices::begin() const noexcept
+{
+  return iterator(0);
+}
+
+_CCCL_NODISCARD inline all_devices::iterator all_devices::end() const noexcept
+{
+  return iterator(static_cast<int>(size()));
+}
+} // namespace __detail
+
+//! @brief A range of all available CUDA devices
+//!
+//! `cuda::devices` provides a view of all available CUDA devices. It is useful for
+//! determining the number of supported devices and for iterating over all devices
+//! in a range-based for loop (e.g., to print device properties, perhaps).
+//!
+//! @par Class synopsis
+//! @code
+//! class __all_devices {                     // exposition only
```

Comment:
This talks about `__all_devices::iterator`

Reply:
i've missed your point.

Reply:
I believe it should be `all_devices::iterator` and not `__all_devices::iterator`

Reply:
I am intentionally not advertising the name of the type. (It's actually defined in the `__detail` namespace.) Users are not supposed to create their own instances of `all_devices`.